include/envoy/ssl/handshaker.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Ssl {
+
+class HandshakerCallbacks {
+public:
+  virtual ~HandshakerCallbacks() = default;
+
+  /**
+   * Hands off the internally-held SSL object for external manipulation.
+   */
+  virtual bssl::UniquePtr<SSL> HandOff() PURE;
+  /**
+   * Accepts an SSL object for internal storage.
+   */
+  virtual void HandBack(bssl::UniquePtr<SSL> ssl) PURE;
+  /**
+   * Called when a handshake is successfully performed.
+   */
+  virtual void OnSuccessCb(SSL* ssl) PURE;
+  /**
+   * Called when a handshake fails.
+   */
+  virtual void OnFailureCb() PURE;
+};
+
+class Handshaker {
+public:
+  virtual ~Handshaker() = default;
+
+  // The initialize method must be called before
+  // doHandshake() is called.
+  virtual void initialize(SSL& ssl) PURE;
+
+  /**
+   * Do the handshake.
+   *
+   *  * |state| is a mutable reference.
+   *  * |ssl| is a pointer, and this method may be called with |ssl| == nullptr,
+   *    for example during a reentrant call to doHandshake() during a period
+   *    when the handshaker has handed off the SSL*.
+   *  * |callbacks| may not exist throughout the lifetime of the Handshaker, and
+   *    should not be stored in an implementation.
+   */
+  virtual Network::PostIoAction doHandshake(SocketState& state, SSL* ssl,
+                                            HandshakerCallbacks& callbacks) PURE;
+
+  /**
+   * Set an internal pointer to the Network::TransportSocketCallbacks struct.
+   * Depending on impl, these callbacks can be invoked to access connection
+   * state, raise connection events, etc.
+   */
+  virtual void setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) PURE;
```

Comment:
Is it possible to defer creation of the handshaker until after the `TransportSocketCallbacks` are set in the owner, so that it can be constructed with them and we don't need this method? (I'm guessing the answer is no, and this is needed, but please double-check).

Reply:
I think the answer is no; while SslSocket::setTransportSocketCallbacks() is called exactly once before handshake operations, it gets called quite late inside ConnectionImpl::ConnectionImpl() (for example), which takes an already-constructed socket as an argument.